README.md
# FormValidator

---

A nice CMD module.

---

## Usage

It is very easy to use this module.

````html
<div class="-FormValidator">
</div>
````

```javascript
seajs.use('FormValidator', function(FormValidator) {
	var validator = new FormValidator('#orderForm',{
		inputs:["transactionPassword"],
		rules: {
			transactionPassword: "required|maxlength:6|digit"
		},
		messages: {
			transactionPassword:{
				required: "交易密码不能为空",
				maxlength : "交易密码为6位数字密码",
				digit:"交易密码只能是数字"
			}
		}
	});
	//表单验证 初始化
	validator.launched();
});
```

## Api

Here is some details.


HISTORY.md
# History

---

## 1.0.0

`new` /FormValidator First version.


tests/FormValidator-spec.js
define(function(require) {

  var FormValidator = require('../src/FormValidator');

  describe('FormValidator', function() {

    it('normal usage', function() {

    });
  });

});


src/FormValidator.js
define(function(c,h,a){var e="BChamFormValidator";var j=['input:not([type]),input[type="color"],input[type="date"],input[type="datetime"],input[type="datetime-local"],input[type="email"],input[type="file"],input[type="hidden"],input[type="month"],input[type="number"],input[type="password"],input[type="range"],input[type="search"],input[type="tel"],input[type="text"],input[type="time"],input[type="url"],input[type="week"],textarea',"select",'input[type="checkbox"],input[type="radio"]'],i=j.join(",");var f={onSubmit:true,onBlur:true,onChange:false,onKeyup:false,onInput:true,wrapper:"div.control-group",success:{},fail:{},rules:{}};var d={required:function(l,k){return k!==null&&$.trim(k).length>0},minlength:function(p,n,o,l){var m=$.trim(n.toString()).length,k=m>=o;if(!l.required){k=k&&m!==0}return k},maxlength:function(l,k,m){return $.trim(k.toString()).length<=m},minvalue:function(m,l,k){return parseFloat(l)>=parseFloat(k)},maxvalue:function(m,k,l){return parseFloat(k)<=parseFloat(l)},regex:function(m,l,k){return k.test(l)},email:function(m,l){var k=/^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i;return k.test($.trim(l))},url:function(m,l){var k=/^(https?|ftp):\/\/(((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:)*@)?(((\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5]))|((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.?)(:\d*)?)(\/((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)+(\/(([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)*)*)?)?(\?((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|[\uE000-\uF8FF]|\/|\?)*)?(\#((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|\/|\?)*)?$/i;return k.test(l)},equals:function(m,k,l){return k===l},ip:function(m,l){var k=/^((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[0-9]{1,2})\.){3}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[0-9]{1,2})$/i;return k.test($.trim(l))},phone:function(m,l){var k=/^(1[^012][0-9]{9})$/i;return k.test($.trim(l))},alpha:function(m,l){var k=/^[a-z]*$/i;return k.test(l)},alpha_numeric:function(m,l){var k=/^[a-z0-9]*$/i;return k.test(l)},alpha_dash:function(m,l){var k=/^[a-z0-9_\-]*$/i;return k.test(l)},digit:function(m,l){var k=/^\d*$/;return k.test(l)},numeric:function(n,m,k){var k=k||null;var l=/^([\+\-]?[0-9]+(\.[0-9]+)?)?$/;if(l.test(m)){if(k==null||m.toString().match(/\./)==null){return true}else{return m.toString().split(".")[1].length<=parseInt(k)}}else{return false}},matches:function(o,n,p,l){var m=$.trim(n).length,k=n===this.fields.filter('[data-id="'+p+'"]').val();if(!l.required){k=k&&m!==0}return k}};var g={required:"输入值不能为空",minlength:"输入值不能少于 :value 个字符",maxlength:"输入值不能超过 :value 个字符",minvalue:"最小值不能超过value",maxvalue:"最大值不能超过value",regex:"请输入正确的值",email:"请输入正确的邮箱格式,如:brucecham@outlook.com",url:"请输入正确的url地址,如:http://www.github.com",equals:"输入值必须等于 :value",ip:"请输入正确的ip地址格式,如:10.1.1.1",phone:"请输入正确的电话号码,如:18688888888",alpha:"输入值只能是a-z之间的小写字母",alpha_numeric:"输入值只能是a-z或0-9",alpha_dash:"输入值只能是a-z,0-9或下划线(_)和(-)",digit:"该项只能输入整数",numeric:"该项只能输入数字",matches:"请再次输入相同的值"};function b(k,l){var n=this;this.form=k;this.options=$.extend({},f,l);this.inputs=this.options.inputs;for(var m=0;m<this.inputs.length;m++){this.inputs[m]="[name="+this.inputs[m]+"]"}this.inputs=this.inputs.toString();this.methods=$.extend({},d);this.fields=k.find(i);this.allowed_rules=[];this.errors={};n._novalidate_old=n.form.attr("novalidate");try{n.form.attr("novalidate","novalidate")}catch(o){}}$.extend(b.prototype,{validate:function(){var k=this;k.fields.filter(k.inputs).each(function(){var l=$(this);l.trigger([e,"validate"].join(":"),[k]);if(k.errors[l.data("id")]){k.renderError.call(k,$(this))}else{k.clearError.call(k,$(this))}})},validateField:function(q,k){var r=$(this),l={},o=r.data("id"),s=null;if(r.is(":disabled")||r.hasClass("disabled")){return}delete k.errors[o];if(r.is('[type="checkbox"], [type="radio"]')){s=r.is(":checked")?r.val():null}else{s=r.val()}var t=k.options.rules[o],p=true;if(t){t=t.split("|");$.each(t,function(w,u){if($.inArray(u,k.allowed_rules)!==-1){l[u]=null}else{var v=u.split(":");if(v.length===2){if($.inArray(v[0],k.allowed_rules)!==-1){l[v[0]]=v[1]}}}});$.each(l,function(v,u){if(k.methods[v].call(k,o,s,u,l)!==true){k.errors[o]=k.format.call(k,o,v,u);p=false;return false}else{delete k.errors[o]}});if(p){var n=k.options.success[o];if(n){n.call(k,r)}}else{var m=k.options.fail[o];if(m){m.call(k,r)}}}$(k).trigger([e,"validate"].join(":"),[r,function(u){k.errors[r.data("id")]=u}])},format:function(l,m,n){var k;if(typeof this.options.messages[l]!=="undefined"&&typeof this.options.messages[l][m]!=="undefined"){k=this.options.messages[l][m]}else{k=g[m]}if($.type(n)!=="undefined"&&n!==null){if($.type(n)==="boolean"||$.type(n)==="string"||$.type(n)==="number"){n={value:n}}$.each(n,function(p,o){k=k.replace(new RegExp(":"+p,"ig"),o)})}return k},clearError:function(l){l=$(l);var k=l.closest("div.controls").children("div.help-block");if(this.options.wrapper!==null){l.closest(this.options.wrapper).removeClass("control-group-error")}k.html("")},renderError:function(l){l=$(l);var k=l.closest("div.controls").children("div.help-block");if(this.options.wrapper!==null){l.closest(this.options.wrapper).addClass("control-group-error")}if(k.length===0){k=$("<div/>",{"class":"help-block"});l.closest("div.controls").children(":last").after(k)}k.html(this.errors[l.data("id")])},clearErrors:function(){},renderErrors:function(){},extendRules:function(k){$.extend(this.methods,k)},parseDom:function(){var k=this;k.fields.each(function(){var l=$(this),m="";if(!l.data("id")){if(l.attr("name")){m=l.attr("name")}else{m=l.attr("id")}l.attr("data-id",m);l.data("id",m)}})},result:function(){var k=this;if(!$.isEmptyObject(k.errors)){return false}k.validate();if(!$.isEmptyObject(k.errors)){return false}else{return true}},launched:function(){var o=this,q=o.options.onSubmit,l=o.options.onBlur,k=o.options.onChange,m=o.options.onKeyup,p=o.options.onInput;o.parseDom();$.each(o.methods,function(s,r){o.allowed_rules.push(s)});o.fields.each(function(){$(this).on([e,"validate"].join(":"),o.validateField)});if(l||k||m||p){var n=[];if(l){n.push("blur")}if(k){n.push("change")}if(m){n.push("keyup")}if(p){n.push("input")}o.fields.filter(j[0]).each(function(){var r=$(this);r.on([n.join(" ")].join("."),function(){r.trigger([e,"validate"].join(":"),[o]);if(o.errors[r.data("id")]){o.renderError.call(o,this)}else{o.clearError.call(o,this)}})})}if(q){o.on("afterValidate",function(s,r){r.find('button[type="submit"]').addClass("disabled");r.find('button[type="submit"]').attr("disabled",true);r.find('button[type="submit"]').text("正在提交，请稍候")});o.form.on(["submit"].join("."),function(r){if(!$.isEmptyObject(o.errors)){return false}o.validate();if(!$.isEmptyObject(o.errors)){return false}else{if(typeof o.options.extendValidator!=="undefined"){if(o.options.extendValidator.call(this)){$(o).trigger([e,"afterValidate"].join(":"),[o.form])}else{return false}}else{$(o).trigger([e,"afterValidate"].join(":"),[o.form])}}})}},destory:function(){},on:function(k,m){var l=this;switch(k){case"validate":$(l).on([e,k].join("."),function(p,o,q){if(!l.errors[o.data("id")]){var n=m(p,o,q);if(n){delete l.errors[o.data("id")]}}});break;case"afterValidate":$(l).on([e,k].join("."),function(o,n){m(o,n)});break}}});a.exports=function(l,k){if(typeof l==="string"){l=$(l)}return new b(l,k)};a.exports.methods=d;a.exports.messages=g});

src/FormValidator-debug.js
/*
//表单验证 引用
var validator = new FormValidator('#orderForm',{
    //验证的input框name
    inputs:["transactionPassword"],
    //对应的验证规则
    rules: {
        transactionPassword: "required|maxlength:6|digit"
    },
    //验证的提示信息
    messages: {
        transactionPassword:{
            required: "交易密码不能为空",
            maxlength : "交易密码为6位数字密码",
            digit:"交易密码只能是数字"
        }
    }
});
//表单验证 初始化
validator.launched();

*/


define(function(require, exports, module) {
    "use strict";
    //使用严格模式
    var namespace = "BChamFormValidator";
    var type = [ 'input:not([type]),input[type="color"],input[type="date"],input[type="datetime"],input[type="datetime-local"],input[type="email"],input[type="file"],input[type="hidden"],input[type="month"],input[type="number"],input[type="password"],input[type="range"],input[type="search"],input[type="tel"],input[type="text"],input[type="time"],input[type="url"],input[type="week"],textarea', "select", 'input[type="checkbox"],input[type="radio"]' ], allTypes = type.join(",");
    var defaults = {
        //默认提交是否验证 false ajax方式提交表单 表单结果从result()取值
        onSubmit: true,
        //全局默认onblur时是否验证
        onBlur: true,
        onChange: false,
        onKeyup: false,
        onInput: true,
        //默认渲染成功和错误样式的dom
        wrapper: "div.control-group",
        success: {},
        fail: {},
        rules: {}
    };
    var methods = {
        required: function(field, value) {
            return value !== null && $.trim(value).length > 0;
        },
        minlength: function(field, value, min_len, all_rules) {
            var length = $.trim(value.toString()).length, result = length >= min_len;
            if (!all_rules["required"]) {
                result = result && length !== 0;
            }
            return result;
        },
        maxlength: function(field, value, max_len) {
            return $.trim(value.toString()).length <= max_len;
        },
        minvalue: function(field, value, min_val){
            return parseFloat(value) >= parseFloat(min_val);
        },
        maxvalue: function(field, value, max_val){
            return parseFloat(value) <= parseFloat(max_val);
        },
        regex: function(field, value, regexp) {
            return regexp.test(value);
        },
        email: function(field, value) {
            var regex = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i;
            return regex.test($.trim(value));
        },
        url: function(field, value) {
            var regex = /^(https?|ftp):\/\/(((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:)*@)?(((\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5]))|((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.?)(:\d*)?)(\/((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)+(\/(([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)*)*)?)?(\?((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|[\uE000-\uF8FF]|\/|\?)*)?(\#((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|\/|\?)*)?$/i;
            return regex.test(value);
        },
        equals: function(field, value, target) {
            return value === target;
        },
        ip: function(field, value) {
            var regex = /^((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[0-9]{1,2})\.){3}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[0-9]{1,2})$/i;
            return regex.test($.trim(value));
        },
        phone: function(field, value) {
            // 调整 注册支持号段：除 100、110、120外都可支持
            var regex = /^(1[^012][0-9]{9})$/i;
            return regex.test($.trim(value));
        },
        alpha: function(field, value) {
            var regex = /^[a-z]*$/i;
            return regex.test(value);
        },
        alpha_numeric: function(field, value) {
            var regex = /^[a-z0-9]*$/i;
            return regex.test(value);
        },
        alpha_dash: function(field, value) {
            var regex = /^[a-z0-9_\-]*$/i;
            return regex.test(value);
        },
        digit: function(field, value) {
            var regex = /^\d*$/;
            return regex.test(value);
        },
        numeric: function(field, value , point_num) {
            var point_num = point_num || null;
            var regex = /^([\+\-]?[0-9]+(\.[0-9]+)?)?$/;
            if( regex.test(value) ){
                if( point_num == null || value.toString().match(/\./) == null){
                    return true;
                }else{
                    return value.toString().split(".")[1].length <= parseInt(point_num);
                }
            }else{
                return false;
            }
        },
        matches: function(field, value, param, all_rules) {
            var length = $.trim(value).length, result = value === this.fields.filter('[data-id="' + param + '"]').val();
            if (!all_rules["required"]) {
                result = result && length !== 0;
            }
            return result;
        }
    };
    var messages = {
        required: "输入值不能为空",
        minlength: "输入值不能少于 :value 个字符",
        maxlength: "输入值不能超过 :value 个字符",
        minvalue: "最小值不能超过value",
        maxvalue: "最大值不能超过value",
        regex: "请输入正确的值",
        email: "请输入正确的邮箱格式,如:brucecham@outlook.com",
        url: "请输入正确的url地址,如:http://www.github.com",
        equals: "输入值必须等于 :value",
        ip: "请输入正确的ip地址格式,如:10.1.1.1",
        phone: "请输入正确的电话号码,如:18688888888",
        alpha: "输入值只能是a-z之间的小写字母",
        alpha_numeric: "输入值只能是a-z或0-9",
        alpha_dash: "输入值只能是a-z,0-9或下划线(_)和(-)",
        digit: "该项只能输入整数",
        numeric: "该项只能输入数字",
        matches: "请再次输入相同的值"
    };
    function Validator($form, options) {
        var validator = this;
        //public
        this.form = $form;
        this.options = $.extend({}, defaults, options);
        this.inputs = this.options.inputs;
        for(var i=0;i<this.inputs.length;i++){
            this.inputs[i] = "[name=" + this.inputs[i] + "]";
        }
        this.inputs = this.inputs.toString();
        this.methods = $.extend({}, methods);
        this.fields = $form.find(allTypes);
        this.allowed_rules = [];
        this.errors = {};
        // 记录 form 原来的 novalidate 的值，因为初始化时需要设置 novalidate 的值，destroy 的时候需要恢复。
        validator._novalidate_old = validator.form.attr("novalidate");
        //禁用html5默认表单验证
        try {
            validator.form.attr("novalidate", "novalidate");
        } catch (e) {}
    }
    $.extend(Validator.prototype, {
        validate: function() {
            var validator = this;
            // validator.fields.filter(type[0]).each(function() {
            validator.fields.filter(validator.inputs).each(function() {
                var $field = $(this);
                $field.trigger([ namespace, "validate" ].join(":"), [ validator ]);
                if (validator.errors[$field.data("id")]) {
                    validator.renderError.call(validator, $(this));
                } else {
                    validator.clearError.call(validator, $(this));
                }
            });
        },
        //scope:$field
        validateField: function(e, validator) {
            var $field = $(this), normalized_rules = {}, field_name = $field.data("id"), value = null;
            // 过滤disabled状态field的逻辑验证
            if ($field.is(":disabled") || $field.hasClass("disabled")) {
                return;
            }
            //清除控件错误提示
            delete validator.errors[field_name];
            if ($field.is('[type="checkbox"], [type="radio"]')) {
                value = $field.is(":checked") ? $field.val() : null;
            } else {
                value = $field.val();
            }
            var rules = validator.options.rules[field_name], rulesResult = true;
            if (rules) {
                rules = rules.split("|");
                $.each(rules, function(rule_idx, rule_value) {
                    if ($.inArray(rule_value, validator.allowed_rules) !== -1) {
                        normalized_rules[rule_value] = null;
                    } else {
                        var rule = rule_value.split(":");
                        if (rule.length === 2) {
                            if ($.inArray(rule[0], validator.allowed_rules) !== -1) {
                                normalized_rules[rule[0]] = rule[1];
                            }
                        }
                    }
                });
                $.each(normalized_rules, function(fn_name, fn_args) {
                    if (validator.methods[fn_name].call(validator, field_name, value, fn_args, normalized_rules) !== true) {
                        validator.errors[field_name] = validator.format.call(validator, field_name, fn_name, fn_args);
                        rulesResult = false;
                        return false;
                    } else {
                        //清除控件错误提示
                        delete validator.errors[field_name];
                    }
                });
                if (rulesResult) {
                    var fieldSuccessFn = validator.options.success[field_name];
                    if (fieldSuccessFn) {
                        fieldSuccessFn.call(validator, $field);
                    }
                } else {
                    var fieldErrorFn = validator.options.fail[field_name];
                    if (fieldErrorFn) {
                        fieldErrorFn.call(validator, $field);
                    }
                }
            }
            $(validator).trigger([ namespace, "validate" ].join(":"), [ $field, function(error) {
                validator.errors[$field.data("id")] = error;
            } ]);
        },
        format: function(field_name, rule, params) {
            var message;
            if (typeof this.options.messages[field_name] !== "undefined" && typeof this.options.messages[field_name][rule] !== "undefined") {
                message = this.options.messages[field_name][rule];
            } else {
                message = messages[rule];
            }
            if ($.type(params) !== "undefined" && params !== null) {
                if ($.type(params) === "boolean" || $.type(params) === "string" || $.type(params) === "number") {
                    params = {
                        value: params
                    };
                }
                $.each(params, function(k, v) {
                    message = message.replace(new RegExp(":" + k, "ig"), v);
                });
            }
            return message;
        },
        clearError: function($field) {
            $field = $($field);
            var $div = $field.closest("div.controls").children("div.help-block");
            if (this.options.wrapper !== null) {
                $field.closest(this.options.wrapper).removeClass("control-group-error");
            }
            $div.html("");
        },
        renderError: function($field) {
            $field = $($field);
            var $div = $field.closest("div.controls").children("div.help-block");
            if (this.options.wrapper !== null) {
                $field.closest(this.options.wrapper).addClass("control-group-error");
            }
            if ($div.length === 0) {
                $div = $("<div/>", {
                    "class": "help-block"
                });
                $field.closest("div.controls").children(":last").after($div);
            }
            $div.html(this.errors[$field.data("id")]);
        },
        clearErrors: function() {},
        renderErrors: function() {},
        extendRules: function(newMethods) {
            $.extend(this.methods, newMethods);
        },
        parseDom: function() {
            var validator = this;
            validator.fields.each(function() {
                var $field = $(this), id = "";
                if (!$field.data("id")) {
                    if ($field.attr("name")) id = $field.attr("name"); else id = $field.attr("id");
                    $field.attr("data-id", id);
                    $field.data("id", id);
                }
            });
        },
        result: function(){
            var validator = this;
            if (!$.isEmptyObject(validator.errors)) {
                return false;
            }
            validator.validate();
            if (!$.isEmptyObject(validator.errors)) {
                return false;
            } else {
                return true;
            }
        },
        launched: function() {
            var validator = this, checkOnSubmit = validator.options.onSubmit, checkOnBlur = validator.options.onBlur, checkOnChange = validator.options.onChange, checkOnKeyup = validator.options.onKeyup ,checkOnInput = validator.options.onInput;
            validator.parseDom();
            $.each(validator.methods, function(k, v) {
                validator.allowed_rules.push(k);
            });
            validator.fields.each(function() {
                $(this).on([ namespace, "validate" ].join(":"), validator.validateField);
            });
            if (checkOnBlur || checkOnChange || checkOnKeyup || checkOnInput) {
                var events = [];
                if (checkOnBlur) events.push("blur");
                if (checkOnChange) events.push("change");
                if (checkOnKeyup) events.push("keyup");
                if(checkOnInput) events.push("input");
                validator.fields.filter(type[0]).each(function() {
                    var $field = $(this);
                    $field.on([ events.join(" ") ].join("."), function() {
                        $field.trigger([ namespace, "validate" ].join(":"), [ validator ]);
                        if (validator.errors[$field.data("id")]) {
                            validator.renderError.call(validator, this);
                        } else {
                            validator.clearError.call(validator, this);
                        }
                    });
                });
            }
            if (checkOnSubmit) {
                validator.on("afterValidate", function(e, $form) {
                    $form.find('button[type="submit"]').addClass("disabled");
                    $form.find('button[type="submit"]').attr("disabled", true);
                    $form.find('button[type="submit"]').text("正在提交，请稍候");
                });
                validator.form.on([ "submit" ].join("."), function(e) {
                    if (!$.isEmptyObject(validator.errors)) {
                        return false;
                    }
                    validator.validate();
                    if (!$.isEmptyObject(validator.errors)) {
                        return false;
                    } else {
                        if(typeof validator.options.extendValidator !== "undefined"){
                            if(validator.options.extendValidator.call(this)){
                                $(validator).trigger([ namespace, "afterValidate" ].join(":"), [ validator.form ]);
                            }else{
                                return false;
                            }
                        }else{
                            $(validator).trigger([ namespace, "afterValidate" ].join(":"), [ validator.form ]);
                        }
                    }
                });
            }
        },
        destory: function() {},
        on: function(eventName, callback) {
            var validator = this;
            switch (eventName) {
              case "validate":
                $(validator).on([ namespace, eventName ].join("."), function(e, $field, errors) {
                    if (!validator.errors[$field.data("id")]) {
                        var result = callback(e, $field, errors);
                        if (result) {
                            delete validator.errors[$field.data("id")];
                        }
                    }
                });
                break;

              case "afterValidate":
                $(validator).on([ namespace, eventName ].join("."), function(e, $form) {
                    callback(e, $form);
                });
                break;
            }
        }
    });
    module.exports = function(form, options) {
        if (typeof form === "string") {
            form = $(form);
        }
        return new Validator(form, options);
    };
    // 暴露出已有的验证方法和提示信息
    module.exports.methods = methods;
    module.exports.messages = messages;
});
